### PR TITLE
Bring up new-style registration API as wrapper around old-style

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#33205 Bring up new-style registration API as wrapper around old-style**

A number of important use-cases are implemented:

- def(schema): defines a schema, with no implementation (alias
  inferred from schema, by default)
- def(schema, fn_ptr): registers fn_ptr as a catch-all kernel
  for the operation
- def(schema, lambda): registers lambda as a catch-all kernel
  for the operation
- def(schema, torch::dispatch(dispatch_key, fn)), and
  def(schema, torch::dispatch(device_type, fn)): registers
  the function to only be executed when dispatch_key/device_type
  is selected for use
- def(schema, CppFunction::makeUnboxedOnly(fn)): registers the function
  as unboxed only, using the inline syntax
- impl(...) variant of def, which will eventually not require full schema string

All of our code generated registrations in ATen are switched to
the new API.

Some aspects of the API which are not fully implemented:

- It's still not valid to omit the schema when registering a function
  pointer, due to #32549
- Although it's possible to take advantage of top-level namespaces
  ala torch::import("aten"), we don't use it because this results
  in worse code (as we have to cat everything back together).  This
  is not an essential problem, we just need the internals to be less
  stupid.

There are some aspects of the API which don't semantically make sense,
but I chose not to fix them in this PR:

- There's no reason to optional<DispatchKey>, DispatchKey would
  work just fine (use DispatchKey::Undefined for the nullopt case)

In the long term, we should swap the wrapper around: the new-style
API has the real implementation, and the old-style API is backwards
compatibility.  However, this implies a lot of internal refactoring,
so I decided to short circuit around it to get this in faster

Ancillary changes:
- I stopped moving optional<DispatchKey>, it's literally just two
  words, pass it by value please.
- Needed to add a & qualified version of RegisterOps::op, since
  I'm storing RegisterOps as a member inside the new style
  Namespace and I cannot conveniently get a rvalue reference
  to it in that situation.  (BTW, register_ = std::move(register_)
  really doesn't work, don't try it!)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19856626](https://our.internmc.facebook.com/intern/diff/D19856626)

### CODE REVIEW

**Comment by ezyang** [comment]
New registrations look like:

```
auto registerer = torch::import()
  .def("aten::add.Tensor(Tensor self, Tensor other, *, Scalar alpha=1) -> Tensor", torch::dispatch(DispatchKey::CPUTensorId, &CPUType::add_Tensor))
  .def("aten::add_.Tensor(Tensor(a!) self, Tensor other, *, Scalar alpha=1) -> Tensor(a!)", torch::dispatch(DispatchKey::CPUTensorId, TORCH_OPTIMIZED_FN(CPUType::add__Tensor)))
  .def("aten::add.out(Tensor self, Tensor other, *, Scalar alpha=1, Tensor(a!) out) -> Tensor(a!)", torch::dispatch(DispatchKey::CPUTensorId, TORCH_OPTIMIZED_FN(CPUType::add_out_out)))
```

File: aten/src/ATen/core/op_registration/op_registration.h
Comment (bwasti) [inline aten/src/ATen/core/op_registration/op_registration.h]: does `torch::import` already exist?
would it be possible to name this `torch::namespace` for clarity?

**Comment by ezyang** [comment]
I spent a while hemming and hawing about how much internal refactoring I would do before implementing this. In the end implementing it as a wrapper was much easier, but I have a number of internal refactors in mind, which could be done before landing this PR, if people so desired.

File: aten/src/ATen/core/op_registration/op_registration.h
Comment (ezyang) [reply aten/src/ATen/core/op_registration/op_registration.h]: No it doesn't, I'm just following @zdevito's original API suggestion. `torch::namespace` is not permissible because namespace is a keyword. We could, alternately, rename `class Namespace` to `class Module`, bringing it more in line with pybind11 nomenclature

Comment (bhosmer) [reply aten/src/ATen/core/op_registration/op_registration.h]: IIRC `import` is already part of TorchBind... the name comes from [pybind](https://pybind11.readthedocs.io/en/stable/advanced/embedding.html#importing-modules) originally.

Comment (bhosmer) [reply aten/src/ATen/core/op_registration/op_registration.h]: IIUC I'd vote for the latter rename - it'd be a shame to lose the familiarity of `import` as used in torchbind/pybind (esp if this will become part of torchbind)

**Review by bhosmer — COMMENTED** [review]
Mod renaming `Namespace` to `Module` and test failures(?) this LGTM, is there any reason we shouldn't land it as-is?

**Comment by ezyang** [comment]
I need reviews on the earlier diffs in the stack, and @smessmer should be given a chance to take a look.

File: aten/src/ATen/core/op_registration/op_registration.h
Comment (smessmer) [inline aten/src/ATen/core/op_registration/op_registration.h]: nit: move constructors and assignment operators should be noexcept whenever possible, otherwise STL data types like vector call copy constructors instead even if moving would be possible.

File: aten/src/ATen/core/op_registration/op_registration.h
Comment (smessmer) [inline aten/src/ATen/core/op_registration/op_registration.h]: add a comment that this can be nullptr for the top level module or use `c10::optional` to encode it explicitly in code.

File: aten/src/ATen/core/op_registration/op_registration.cpp
Comment (smessmer) [inline aten/src/ATen/core/op_registration/op_registration.cpp]: The difference in link time should be minimal, and this probably costs you an (also minimal) performance hit.

File: aten/src/ATen/core/op_registration/op_registration.cpp
Comment (smessmer) [inline aten/src/ATen/core/op_registration/op_registration.cpp]: Shouldn't we set `AliasAnalysisKind::FROM_SCHEMA` here?

File: aten/src/ATen/core/op_registration/op_registration.cpp
Comment (smessmer) [inline aten/src/ATen/core/op_registration/op_registration.cpp]: I'm a bit confused here. How do you handle `.kernel()` vs `.catchAllKernel()` vs `.impl_unboxedOnlyKernel()` vs `.impl_unboxedOnlyCatchAllKernel()`? I think there's some logic missing here.

**Review by smessmer — CHANGES_REQUESTED** [review]
requesting changes because of the missing logic to distinguish kernel/catchAllKernel/impl_...

File: aten/src/ATen/core/op_registration/op_registration.cpp
Comment (ezyang) [reply aten/src/ATen/core/op_registration/op_registration.cpp]: So, in a few versions of my diff, I managed to fuck up my linker so that it took like 10G of memory and several minutes to link `libtorch_cpu.so`. I never figured exactly what was causing it though... :/

File: aten/src/ATen/core/op_registration/op_registration.cpp
Comment (ezyang) [reply aten/src/ATen/core/op_registration/op_registration.cpp]: This is related to the design discussion we had on only having `def`, versus having two variants, `def` and `override`. Basil can fill you in.

File: aten/src/ATen/core/op_registration/op_registration.cpp
Comment (smessmer) [reply aten/src/ATen/core/op_registration/op_registration.cpp]: oh nvm, you're calling the private `.kernel()` API here, that means CppFunction should handle this distinction, not the code here. I still think it's missing though.

File: aten/src/ATen/core/op_registration/op_registration.cpp
Comment (ezyang) [reply aten/src/ATen/core/op_registration/op_registration.cpp]: I'll go ahead and remove this, as I haven't really established a causal relationship here.

File: aten/src/ATen/core/op_registration/op_registration.cpp
Comment (ezyang) [reply aten/src/ATen/core/op_registration/op_registration.cpp]: The answer is that it is handled by `TORCH_OPTIMIZED_FN`. So there are two constructors in CppFunction:

```
    // This overload accepts function pointers, e.g., CppFunction(&add_impl)
    template <typename Func>
    explicit CppFunction(Func* f, std::enable_if_t<guts::is_function_type<Func>::value, std::nullptr_t> = nullptr)
      : func_(c10::KernelFunction::makeFromUnboxedRuntimeFunction(f))
      // TODO: Don't go through WrapRuntimeKernelFunctor
      , schema_(detail::FunctionSchemaInferer<detail::WrapRuntimeKernelFunctor<std::decay_t<Func>>>()())
      {}  

```

That's `kernel` and `catchAllKernel`

Then there's

```
    template <typename Func, Func* f>
    static CppFunction makeOptimized() {
      return CppFunction(
        // TODO: For some reason, the "optimized" codepath is still using the
        // RuntimeFunction functor, which is kind of weird
        c10::KernelFunction::makeFromUnboxedOnlyRuntimeFunction(f),
        // NB: disable function schema inference because some ops don't support it yet (TODO fix)
        /* schema */ nullptr
      );
    }
```

That's unboxedOnly

Comment (ezyang) [reply aten/src/ATen/core/op_registration/op_registration.cpp]: Some of the confusion is because I got the names wrong. I'm going to rename `TORCH_OPTIMIZED_FN` to just `CppFunction::makeUnboxedOnly` for clarity.

**Comment by ezyang** [comment]
@smessmer all your comments addressed

File: aten/src/ATen/core/op_registration/op_registration.h
Comment (bhosmer) [inline aten/src/ATen/core/op_registration/op_registration.h]: quick nit: s/override/impl/

File: aten/src/ATen/core/op_registration/op_registration.h
Comment (gchanan) [inline aten/src/ATen/core/op_registration/op_registration.h]: nit: I think "with multiple implementations (one per backend)" is clearer.

File: aten/src/ATen/core/op_registration/op_registration.cpp
Comment (smessmer) [inline aten/src/ATen/core/op_registration/op_registration.cpp]: Using rvalue overloads and returning && is a way of pushing people towards using method chaining as in

    torch::import().def().def().def();

If you want to allow non-chained calls, i.e.

    auto m = torch::import().def();
    m.def();
    m.def();

then you might need to implement all those function twice, once with an rvalue overload that returns an rvalue (or a value) and once with an lvalue overload that returns a lvalue (this second overloads return can't be assigned to anything but it can still be used for chaining).

I guess it depends on what pybind does here and we should follow its api.

File: aten/src/ATen/core/op_registration/op_registration.cpp
Comment (smessmer) [inline aten/src/ATen/core/op_registration/op_registration.cpp]: Anonymous namespaces are more readable than the much overloaded static keyword and do the same thing here because this is not a method. You could also make it a static and private method on the module class if you're never calling it from anywhere else.

File: aten/src/ATen/core/op_registration/op_registration.cpp
Comment (smessmer) [inline aten/src/ATen/core/op_registration/op_registration.cpp]: I think calling .def() multiple times for the same operator should be a hard error. The current implementation allows it and merges.

**Review by smessmer — DISMISSED** [review]
Thanks for clearing up the unboxedonly thing, makes sense.

PR mostly looks good. What I'd like to see added here are loads and loads of unit tests, especially for use cases that we don't cover with our internal ops.

File: aten/src/ATen/core/op_registration/op_registration.cpp
Comment (smessmer) [inline aten/src/ATen/core/op_registration/op_registration.cpp]: Basil just noticed that the argument here should be called "schema_or_unqual_name" instead of "unqual_name"

File: aten/src/ATen/core/op_registration/op_registration.cpp
Comment (ezyang) [reply aten/src/ATen/core/op_registration/op_registration.cpp]: pybind11 doesn't do chaining. Instead, the PYBIND11_MODULE macro brings an `m` into scope, and then people just call it like a normal method. There is no chaining support.

File: aten/src/ATen/core/op_registration/op_registration.cpp
Comment (ezyang) [reply aten/src/ATen/core/op_registration/op_registration.cpp]: I agree, but this increases the scope of the PR substantially, as I now have to start modifying how the existing RegisterOperators mechanism works to enforce this behavior. My preference is for this issue to resolve itself "automatically" when schema merging gets implemented, in a separate PR. If you like, we can block this PR until schema merging lands, but in this case, the impermissible behavior is very mild and I think it's OK to land this first, and schema merging later. What do you think?

**Comment by ezyang** [comment]
Re testing: we've had disagreements about what the best way to go about testing these APIs are.

Here is what I suggest: we start porting over the old "existing" tests to use the new API. However, I don't think it makes sense to force the tests into the same PR, as I would also like to, at the same time, make the existing c10 test suite more concise, and repeat itself less. I think the existing "coverage" (where internal registrations are moved over to use this API) is sufficient testing, and as a separate workstream we should figure out what to do with the existing test suite.

File: aten/src/ATen/core/op_registration/op_registration.cpp
Comment (smessmer) [reply aten/src/ATen/core/op_registration/op_registration.cpp]: wait, but in pybind, you can do `...def().def()` right?

For `m.def()` we need at least the lvalue overload here

File: aten/src/ATen/core/op_registration/op_registration.cpp
Comment (bhosmer) [reply aten/src/ATen/core/op_registration/op_registration.cpp]: Well, if this is truly `schema_or_unqual_name` then our prepending of namespaces is inconsistent - in `def(schema)` we're passing it through unmodified.

File: aten/src/ATen/core/op_registration/op_registration.cpp
Comment (ezyang) [reply aten/src/ATen/core/op_registration/op_registration.cpp]: I think for now I won't add the lvalue overloads but we should talk about how "close to pybind11" we actually want the API to be next meeting.

File: aten/src/ATen/core/op_registration/op_registration.cpp
Comment (bhosmer) [reply aten/src/ATen/core/op_registration/op_registration.cpp]: Adding this as a review so the preceding doesn't get lost - seems like we need to either
a) handle names and schemas uniformly w.r.t. prepending namespaces, or 
b) snoop the string to decide whether it's a name or schema, or
c) define this version of `def` as taking unqualified names and never schemas as its first argument

Comment (ezyang) [reply aten/src/ATen/core/op_registration/op_registration.cpp]: Yes there's a bug. It wasn't caught in the integration tests because the `impl()` all have full schema so you end up with the "right" thing in the end.

**Comment by ezyang** [comment]
@smessmer Ready!

File: aten/src/ATen/core/op_registration/op_registration.cpp
Comment (bhosmer) [inline aten/src/ATen/core/op_registration/op_registration.cpp]: FWIW if we're not actually going to insist that `unqual_name_or_schema` be unqualified we should probably rename it (here and elsewhere). 

Also in that case there's namespace-chaining behavior that we might want to describe in the "New style API" usage comment.

Comment (bhosmer) [reply aten/src/ATen/core/op_registration/op_registration.cpp]: BTW curious why this is a function than a private method that just adds `ns_` directly...

Comment (ezyang) [reply aten/src/ATen/core/op_registration/op_registration.cpp]: I guess I can just call it `name_or_schema`. As for why it's a function, it's so that I don't have to declare it as a method in the header file. That's a big win, at least in my book :)

File: aten/src/ATen/core/op_registration/op_registration.cpp
Comment (smessmer) [reply aten/src/ATen/core/op_registration/op_registration.cpp]: You could hack it together by just remembering a global (or static within Module) std::vector of all ops that got registered with `.def()`. That should be relatively small scope and not require changing the RegisterOperators mechanism.

I think we should do that before landing this to avoid people starting to use the impermissible behavior and running into bc issues once we remove it.

Comment (ezyang) [reply aten/src/ATen/core/op_registration/op_registration.cpp]: But this is all wasted code when schema merging gets written. Backwards compatibility is not a do or die thing; you can (and should!) declare APIs to be unstable while you're still working on them (which is the case here!), otherwise you are just making a lot of busy work for yourself, without any benefit!

Comment (ezyang) [reply aten/src/ATen/core/op_registration/op_registration.cpp]: One thing I can do is file an issue describing all of the "over-permissiveness" we need to audit and test for before moving the API out of beta.

**Review by smessmer — APPROVED** [review]
Still a bit hesitant about test coverage, but accepting to unblock. Feel free to land and add tests later.

**Review by bhosmer — APPROVED** [review]
Yeah let's circle back if we want to beef up the tests

**Pull request closed without merging**
